### PR TITLE
W-11659429: Adopt AMF 5

### PR DESCRIPTION
**Pull request by juanchib**

### CODE REVIEW

File: interface-impl-amf/src/main/java/org/mule/amf/impl/model/YamlParameterValidationStrategy.java [hunk 1]
```diff
@@ -6,24 +6,32 @@
...
-    super(anyShape, schemaNeedsQuotes);
+  private final LazyValue<AMFShapePayloadValidator> parameterValidator =
+      new LazyValue<>(() -> getAmfConfiguration().elementClient().payloadValidatorFor(anyShape, APPLICATION_YAML,
+                                                                                      ValidationMode
+                                                                                          .ScalarRelaxedValidationMode()));
+
+  public YamlParameterValidationStrategy(AnyShape anyShape, boolean schemaNeedsQuotes, AMFConfiguration amfConfiguration) {
+    super(schemaNeedsQuotes);
+    this.anyShape = anyShape;
+    this.amfConfiguration = amfConfiguration;
```
Comment (dcourtil) [inline interface-impl-amf/src/main/java/org/mule/amf/impl/model/YamlParameterValidationStrategy.java:34]:
```suggestion
    this.parameterValidator =
      new LazyValue<>(() -> amfConfiguration.elementClient().payloadValidatorFor(anyShape, APPLICATION_YAML, ValidationMode.ScalarRelaxedValidationMode()));
```

File: interface-impl-amf/src/main/java/org/mule/amf/impl/parser/factory/AMFParserWrapper.java [hunk 2]
```diff
@@ -6,46 +6,76 @@
...
+      final String rootDir = file.isDirectory() ? file.getPath() : file.getParent();
+      amfConfiguration = amfConfiguration.withResourceLoader(new ExchangeDependencyResourceLoader(rootDir, execEnv));
+    }
+
+    AMFParseResult amfParseResult = handleFuture(amfConfiguration.baseUnitClient()
+        .parse(URLDecoder.decode(apiRef.getPathAsUri().toString())));
+    this.parsingIssues = amfParseResult.results();
+    this.model = amfParseResult.baseUnit();
+    this.spec = amfParseResult.sourceSpec();
+    this.apiConfiguration = APIConfiguration.fromSpec(spec).withExecutionEnvironment(execEnv);
```
Comment (dcourtil) [inline interface-impl-amf/src/main/java/org/mule/amf/impl/parser/factory/AMFParserWrapper.java:66]:
```suggestion
    this.apiConfiguration = APIConfiguration.fromSpec(amfParseResult.sourceSpec()).withExecutionEnvironment(execEnv);
```

File: interface-impl-amf/src/main/java/org/mule/amf/impl/parser/factory/AMFParserWrapper.java [hunk 3]
```diff
@@ -6,46 +6,76 @@
...
+    return (Document) client.transform(model, PipelineId.Editing()).baseUnit();
   }
 
-  public ValidationReport getParsingReport(Document resolvedDoc) throws ParserException {
-    return handleFuture(AMF.validateResolved(resolvedDoc, profileName, messageStyle, environment));
+
+  public AMFValidationReport getParsingReport(Document resolvedDoc) throws ParserException {
+    AMFValidationReport report = handleFuture(client.validate(resolvedDoc));
+
+    return report;
```
Comment (dcourtil) [inline interface-impl-amf/src/main/java/org/mule/amf/impl/parser/factory/AMFParserWrapper.java:78]:
```suggestion
    return handleFuture(client.validate(resolvedDoc));
```

File: parser-service/pom.xml [hunk 4]
```diff
@@ -40,12 +40,6 @@
             <groupId>org.mule.apikit</groupId>
             <artifactId>raml-parser-interface-impl-amf</artifactId>
             <version>${project.version}</version>
-            <exclusions>
```
Comment (dcourtil) [inline parser-service/pom.xml:43]: Why is this not required anymore?

Comment (juanchib) [reply parser-service/pom.xml:43]: Is fixed on AMF 5 we had [a work](https://gus.lightning.force.com/lightning/r/a07EE00000wrF3qYAE/view) about this

File: interface-impl-amf/src/test/java/org/mule/amf/impl/parser/rule/ApiValidationResultImplTest.java [hunk 5]
```diff
@@ -121,21 +120,21 @@ public void testEncodedMessage() throws Exception {
...
     assertEquals(actualMessage, expectedMessage);
   }
 
-  public ValidationResult createAMFValidationResult(String message, Position startPosition, Option<String> location) {
-    Option<LexicalInformation> position = Option.apply(new LexicalInformation(new Range(startPosition, new Position(0, 0))));
-    AMFValidationResult amfValidationResult =
-        new AMFValidationResult(message, "level", "targetNode", Option.empty(), "validationId", position, location, "");
-    return new ValidationResult(amfValidationResult);
+  public AMFValidationResult createAMFValidationResult(String message, Position startPosition, Option<String> location) {
+    Option<LexicalInformation> position =
```
Comment (dcourtil) [inline interface-impl-amf/src/test/java/org/mule/amf/impl/parser/rule/ApiValidationResultImplTest.java:134]: Position is never used

File: interface-impl-amf/src/main/java/org/mule/amf/impl/model/MimeTypeImpl.java [hunk 6]
```diff
@@ -6,17 +6,21 @@
...
+import amf.apicontract.client.platform.AMFConfiguration;
+import amf.apicontract.client.platform.model.domain.Encoding;
+import amf.apicontract.client.platform.model.domain.Payload;
+import amf.core.client.common.validation.ValidationMode;
+import amf.core.client.platform.model.domain.PropertyShape;
+import amf.core.client.platform.model.domain.Shape;
+import amf.core.client.platform.validation.AMFValidationReport;
+import amf.core.client.platform.validation.payload.AMFShapePayloadValidator;
+import amf.shapes.client.platform.ShapesConfiguration;
+import amf.shapes.client.platform.ShapesElementClient;
```
Comment (dcourtil) [inline interface-impl-amf/src/main/java/org/mule/amf/impl/model/MimeTypeImpl.java:18]: Not used anymore

File: pom.xml [hunk 7]
```diff
@@ -73,6 +73,12 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
+            <groupId>org.mule.runtime</groupId>
+            <artifactId>mule-module-logging</artifactId>
```
Comment (dcourtil) [inline pom.xml:78]: What's the purpose of this one?

File: interface-impl-amf/src/main/java/org/mule/amf/impl/model/JsonParameterValidationStrategy.java [hunk 8]
```diff
@@ -6,31 +6,35 @@
...
 
 import static org.mule.amf.impl.model.MediaType.APPLICATION_JSON;
 import static org.mule.amf.impl.model.MediaType.APPLICATION_YAML;
 
 class JsonParameterValidationStrategy extends ValidationStrategy {
 
-  private final LazyValue<PayloadValidator> jsonValidator =
-      new LazyValue<>(() -> schema.payloadValidator(APPLICATION_JSON)
-          .orElseThrow(() -> new ParserException(APPLICATION_JSON + " validator not found for shape " + schema)));
+  private AMFElementClient client;
```
Comment (iglosiggio) [inline interface-impl-amf/src/main/java/org/mule/amf/impl/model/JsonParameterValidationStrategy.java:22]: Do we need to keep a reference to the client here or only on the LazyValue (like on the YamlParameterValidationStrategy).

File: interface-impl-amf/src/main/java/org/mule/amf/impl/model/ParameterValidationStrategyFactory.java [hunk 9]
```diff
@@ -17,14 +18,16 @@ private ParameterValidationStrategyFactory() {
     throw new IllegalStateException("Utility class");
   }
 
-  static ParameterValidationStrategy getStrategy(AnyShape anyShape, boolean schemaNeedsQuotes) {
-    return isYamlValidationNeeded(anyShape) ? new YamlParameterValidationStrategy(anyShape, schemaNeedsQuotes)
-        : getJsonParameterValidationStrategy(anyShape, schemaNeedsQuotes);
+  static ParameterValidationStrategy getStrategy(AnyShape anyShape, boolean needsCharEscaping,
+                                                 AMFConfiguration amfConfiguration) {
+    return isYamlValidationNeeded(anyShape) ? new YamlParameterValidationStrategy(anyShape, needsCharEscaping, amfConfiguration)
```
Comment (iglosiggio) [inline interface-impl-amf/src/main/java/org/mule/amf/impl/model/ParameterValidationStrategyFactory.java:23]: I would like for both the Yaml and JSON strategy to receive either the element client OR the configuration object. It looks weird for it to have different parameterization.

File: interface-impl-amf/src/main/java/org/mule/amf/impl/parser/rule/ApiValidationResultImpl.java [hunk 10]
```diff
@@ -56,10 +57,11 @@ public String getPath() {
 
   @Override
   public Severity getSeverity() {
-    return !severities.contains(validationResult.level().toUpperCase()) ? ERROR : Severity.fromString(validationResult.level());
+    return !severities.contains(validationResult.severityLevel().toUpperCase()) ? ERROR
+        : Severity.fromString(validationResult.severityLevel());
   }
 
-  private static String getPositionMessage(Position startPosition) {
+  private String getPositionMessage(Position startPosition) {
```
Comment (iglosiggio) [inline interface-impl-amf/src/main/java/org/mule/amf/impl/parser/rule/ApiValidationResultImpl.java:64]: Why it's not static anymore?

File: pom.xml (hunk 7 above)
Comment (juanchib) [reply pom.xml:78]: We init a couple of loggers but we never use them,, I will remove it to keep the code clean,, we can add it again if needed

File: interface-impl-amf/src/main/java/org/mule/amf/impl/AMFParser.java [hunk 11]
```diff
@@ -99,34 +101,22 @@ public WebApi getWebApi() {
```
Comment (JuanAller) [inline interface-impl-amf/src/main/java/org/mule/amf/impl/AMFParser.java:101]: getWebApi not used

File: interface-impl-amf/src/main/java/org/mule/amf/impl/parser/factory/AMFParserWrapper.java [hunk 12]
```diff
@@ -6,46 +6,69 @@
...
+    AMFConfiguration amfConfiguration = APIConfiguration
+        .API()
+        .withExecutionEnvironment(execEnv);
+
+    if (apiRef.getResourceLoader().isPresent()) {
+      amfConfiguration = amfConfiguration.withResourceLoader(new ProvidedResourceLoader(apiRef.getResourceLoader().get()));
+    }
+    URI apiUri = apiRef.getPathAsUri();
+
+    if (apiUri.getScheme() != null && apiUri.getScheme().startsWith("file")) {
```
Comment (JuanAller) [inline interface-impl-amf/src/main/java/org/mule/amf/impl/parser/factory/AMFParserWrapper.java:51]: why needed now ?

File: pom.xml [hunk 13]
```diff
@@ -158,7 +158,7 @@
                         <id>apply-format</id>
                         <phase>compile</phase>
                         <goals>
-                            <goal>validate</goal>
+                            <goal>format</goal>
```
Comment (JuanAller) [inline pom.xml:161]: revert this to validate

**Review by JuanAller — DISMISSED** [review]

File: interface-impl-amf/src/main/java/org/mule/amf/impl/AMFParser.java (hunk 11 above)
Comment (juanchib) [reply interface-impl-amf/src/main/java/org/mule/amf/impl/AMFParser.java:101]: This is used at the parser service

File: interface-impl-amf/src/main/java/org/mule/amf/impl/parser/factory/AMFParserWrapper.java (hunk 12 above)
Comment (juanchib) [reply interface-impl-amf/src/main/java/org/mule/amf/impl/parser/factory/AMFParserWrapper.java:51]: This was also needed before to add the exchange resource loader but was moved.

**Review by dcourtil — APPROVED** [review]

**Pull request merged**